### PR TITLE
T: introduce RsConsoleCompletionTestFixture

### PR DESCRIPTION
**Pull request by Undin**
Refactor console completion tests to reuse code from completion test fixtures and avoid code duplication

### CODE REVIEW

File: src/test/kotlin/org/rust/ide/console/RsConsoleCompletionTestFixture.kt [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+        // create main.rs file, which will be used as crate root
+        InlineFile(myFixture, "", "main.rs")
+
+        InlineFile(myFixture, lastLine.trimIndent(), RsConsoleView.VIRTUAL_FILE_NAME).withCaret()
+
+        val codeFragment = myFixture.file as RsReplCodeFragment
+        val crateRoot = codeFragment.crateRoot as RsFile?
+        codeFragment.context = RsConsoleCodeFragmentContext.createContext(myFixture.project, crateRoot, previousLines.trimIndent())
+        BasePlatformTestCase.assertNotNull(codeFragment.crateRoot)
```
Comment (dima74) [inline src/test/kotlin/org/rust/ide/console/RsConsoleCompletionTestFixture.kt:29]:
```suggestion
        assertNotNull(codeFragment.crateRoot)
```

File: src/test/kotlin/org/rust/lang/core/completion/RsCompletionTestFixture.kt [hunk 2]
```diff
@@ -5,125 +5,15 @@
...
-                return
-            }
-            fun LookupElement.debug(): String = "$lookupString ($psiElement)"
-            error("Expected a single completion, but got ${variants.size}\n"
-                + variants.joinToString("\n") { it.debug() })
-        }
-    }
-
-    fun doFirstCompletion(before: String, after: String) {
-        check(hasCaretMarker(before) && hasCaretMarker(after)) {
```
Comment (dima74) [inline src/test/kotlin/org/rust/lang/core/completion/RsCompletionTestFixture.kt:41]: Maybe keep `hasCaretMarker(after)` check here and in `doSingleCompletion`? Or add such check to `checkByText`

File: src/test/kotlin/org/rust/lang/core/completion/RsCompletionTestFixtureBase.kt [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+import org.rust.*
+
+abstract class RsCompletionTestFixtureBase<IN>(
+    protected val myFixture: CodeInsightTestFixture
+) : BaseFixture() {
+
+    private val project: Project get() = myFixture.project
+
+    fun executeSoloCompletion() {
+        val variants = myFixture.completeBasic()
```
Comment (dima74) [inline src/test/kotlin/org/rust/lang/core/completion/RsCompletionTestFixtureBase.kt:23]: Maybe rename `variants` to `lookups` (here and in other functions) for consistency with `checkContainsCompletion`

File: src/test/kotlin/org/rust/lang/core/completion/RsCompletionTestFixtureBase.kt [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    fun checkContainsCompletion(code: IN, variants: List<String>) {
+        prepare(code)
+        val lookups = myFixture.completeBasic()
+
+        checkNotNull(lookups) {
+            "Expected completions that contain $variants, but no completions found"
+        }
+        for (variant in variants) {
+            if (lookups.all { it.lookupString != variant }) {
+                error("Expected completions that contain $variant, but got ${lookups.toList()}")
```
Comment (dima74) [inline src/test/kotlin/org/rust/lang/core/completion/RsCompletionTestFixtureBase.kt:98]: Maybe replace `${lookups.toList()}` to `${lookups.map { it.lookupString }}`

File: src/test/kotlin/org/rust/lang/core/completion/RsCompletionTestFixture.kt (hunk 2 above)
Comment (Undin) [reply src/test/kotlin/org/rust/lang/core/completion/RsCompletionTestFixture.kt:41]: This check is performed in `org.rust.InlineFile#withCaret` that we call in `org.rust.lang.core.completion.RsCompletionTestFixture#prepare`

Comment (dima74) [reply src/test/kotlin/org/rust/lang/core/completion/RsCompletionTestFixture.kt:41]: Check in `org.rust.InlineFile#withCaret` is performed for `before`, and for `after` there is no such check if I understand correctly

Comment (Undin) [reply src/test/kotlin/org/rust/lang/core/completion/RsCompletionTestFixture.kt:41]: You are right
Thanks!

**Comment by Undin** [comment]
bors r=dima74

**Pull request merged**